Index: xc3sprog-r742/CMakeLists.txt
===================================================================
--- xc3sprog-r742.orig/CMakeLists.txt	2014-09-23 17:17:42.507053279 +0800
+++ xc3sprog-r742/CMakeLists.txt	2014-09-23 17:22:59.147040108 +0800
@@ -30,13 +30,7 @@
 if(EXISTS ${PROJECT_SOURCE_DIR}/.git) 
   set(VERSION_STRING "${xc3sprog_VERSION_MAJOR}.${xc3sprog_VERSION_MINOR}-git")
 else(EXISTS ${PROJECT_SOURCE_DIR}/.git)
-  FIND_PACKAGE(Subversion)
-  IF(Subversion_FOUND)
-    Subversion_WC_INFO(${PROJECT_SOURCE_DIR} Project)
-    set(VERSION_STRING "${xc3sprog_VERSION_MAJOR}.${xc3sprog_VERSION_MINOR}-svn${Project_WC_REVISION}")
-  ELSE(Subversion_FOUND)
     set(VERSION_STRING "unknown")
-  ENDIF(Subversion_FOUND)
 endif(EXISTS ${PROJECT_SOURCE_DIR}/.git)
 
 # Create suffix to eventually install inlib64
