diff --git a/api-example.c b/api-example.c
index 639ffc5..5eec66c 100644
--- a/api-example.c
+++ b/api-example.c
@@ -1,5 +1,6 @@
 /*
  * Copyright 2011 Kano
+ * Copyright 2014 Mikeqin
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the Free
@@ -13,6 +14,7 @@
 
 #include "config.h"
 
+#include <assert.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -36,7 +38,7 @@
 	#define INVSOCK -1
 	#define CLOSESOCKET close
 
-	#define SOCKETINIT {}
+	#define SOCKETINIT do{}while(0)
 
 	#define SOCKERRMSG strerror(errno)
 #endif
@@ -110,7 +112,6 @@
 
 	static char *WSAErrorMsg()
 	{
-		char *msg;
 		int i;
 		int id = WSAGetLastError();
 
@@ -128,17 +129,22 @@
 
 	static WSADATA WSA_Data;
 
-	#define SOCKETINIT	int wsa; \
-				if (wsa = WSAStartup(0x0202, &WSA_Data)) { \
-					printf("Socket startup failed: %d\n", wsa); \
-					return 1; \
-				}
+	#define SOCKETINIT	do {  \
+				    int wsa; \
+				    if (wsa = WSAStartup(0x0202, &WSA_Data)) { \
+					    printf("Socket startup failed: %d\n", wsa); \
+					    return 1; \
+					}  \
+				} while (0)
 
 	#ifndef SHUT_RDWR
 	#define SHUT_RDWR SD_BOTH
 	#endif
 #endif
 
+#undef RECVSIZE
+#define RECVSIZE 65500
+
 static const char SEPARATOR = '|';
 static const char COMMA = ',';
 static const char EQ = '=';
@@ -189,16 +195,22 @@ void display(char *buf)
 
 int callapi(char *command, char *host, short int port)
 {
-	char buf[RECVSIZE+1];
+	size_t bufsz = RECVSIZE;
+	char *buf = malloc(bufsz + 1);
 	struct hostent *ip;
 	struct sockaddr_in serv;
 	SOCKETTYPE sock;
 	int ret = 0;
 	int n, p;
 
+	assert(buf);
 	SOCKETINIT;
 
 	ip = gethostbyname(host);
+	if (!ip) {
+		printf("Failed to resolve host %s\n", host);
+		return 1;
+	}
 
 	sock = socket(AF_INET, SOCK_STREAM, 0);
 	if (sock == INVSOCK) {
@@ -224,8 +236,14 @@ int callapi(char *command, char *host, short int port)
 	else {
 		p = 0;
 		buf[0] = '\0';
-		while (p < RECVSIZE) {
-			n = recv(sock, &buf[p], RECVSIZE - p , 0);
+		while (1) {
+			if (bufsz < RECVSIZE + p) {
+				bufsz *= 2;
+				buf = realloc(buf, bufsz);
+				assert(buf);
+			}
+
+			n = recv(sock, &buf[p], RECVSIZE, 0);
 
 			if (SOCKETFAIL(n)) {
 				printf("Recv failed: %s\n", SOCKERRMSG);
@@ -240,14 +258,18 @@ int callapi(char *command, char *host, short int port)
 			buf[p] = '\0';
 		}
 
-		if (ONLY)
-			printf("%s\n", buf);
-		else {
+		if (!ONLY)
 			printf("Reply was '%s'\n", buf);
+		else
+			printf("%s\n", buf);
+
+		if (!ONLY)
 			display(buf);
-		}
 	}
 
+	if (buf)
+	    free(buf);
+
 	CLOSESOCKET(sock);
 
 	return ret;
@@ -281,7 +303,7 @@ int main(int argc, char *argv[])
 		if (strcmp(argv[1], "-?") == 0
 		||  strcmp(argv[1], "-h") == 0
 		||  strcmp(argv[1], "--help") == 0) {
-			fprintf(stderr, "usAge: %s [command [ip/host [port]]]\n", argv[0]);
+			fprintf(stderr, "Usage: %s [command [ip/host [port]]]\n", argv[0]);
 			return 1;
 		}
 
diff --git a/cgminer.c b/cgminer.c
index 3e20851..8cb9d2b 100644
--- a/cgminer.c
+++ b/cgminer.c
@@ -1136,10 +1136,10 @@ static struct opt_table opt_config_table[] = {
 #ifdef USE_AVALON2
 	OPT_WITH_CBARG("--avalon2-freq",
 		     set_avalon2_freq, NULL, &opt_set_avalon2_freq,
-		     "Set frequency range for Avalon2, single value or range, step: 25"),
+		     "Set frequency range for Avalon2, single value or range"),
 	OPT_WITH_CBARG("--avalon2-voltage",
 		     set_avalon2_voltage, NULL, &opt_set_avalon2_voltage,
-		     "Set Avalon2 core voltage, in millivolts, step: 125"),
+		     "Set Avalon2 core voltage, in millivolts"),
 	OPT_WITH_CBARG("--avalon2-fan",
 		     set_avalon2_fan, NULL, &opt_set_avalon2_fan,
 		     "Set Avalon2 target fan speed"),
diff --git a/driver-avalon2.c b/driver-avalon2.c
index 2cdb35a..13bbfef 100644
--- a/driver-avalon2.c
+++ b/driver-avalon2.c
@@ -239,7 +239,7 @@ static void adjust_fan(struct avalon2_info *info)
 	int t;
 
 	if (opt_avalon2_fan_fixed == FAN_FIXED) {
-		info->fan_pct = AVA2_DEFAULT_FAN_PWM;
+		info->fan_pct = opt_avalon2_fan_min;
 		info->fan_pwm = get_fan_pwm(info->fan_pct);
 		return;
 	}
@@ -278,6 +278,20 @@ static inline int mm_cmp_1406(struct avalon2_info *info)
 	return 0;
 }
 
+static inline int mm_cmp_1409(struct avalon2_info *info)
+{
+	/* <= 1409 return 1 */
+	char *mm_1409 = "1409";
+	int i;
+	for (i = 0; i < AVA2_DEFAULT_MODULARS; i++) {
+		if (info->enable[i] &&
+		    strncmp(info->mm_version[i] + 2, mm_1409, 4) <= 0)
+			return 1;
+	}
+
+	return 0;
+}
+
 static int decode_pkg(struct thr_info *thr, struct avalon2_ret *ar, uint8_t *pkg)
 {
 	struct cgpu_info *avalon2 = thr->cgpu;
@@ -483,6 +497,7 @@ static int avalon2_send_pkgs(struct cgpu_info *avalon2, const struct avalon2_pkg
 
 static void avalon2_stratum_pkgs(struct cgpu_info *avalon2, struct pool *pool)
 {
+	struct avalon2_info *info = avalon2->device_data;
 	const int merkle_offset = 36;
 	struct avalon2_pkg pkg;
 	int i, a, b, tmp;
@@ -550,6 +565,16 @@ static void avalon2_stratum_pkgs(struct cgpu_info *avalon2, struct pool *pool)
 	if (pool->coinbase_len > AVA2_P_COINBASE_SIZE) {
 		int coinbase_len_posthash, coinbase_len_prehash;
 		uint8_t coinbase_prehash[32];
+
+	if (mm_cmp_1409(info)) {
+		applog(LOG_INFO, "Avalon2: MM version less then 1409");
+	} else {
+		memset(pkg.data, 0, AVA2_P_DATA_LEN);
+		avalon2_init_pkg(&pkg, AVA2_P_LONGCOINBASE, 1, 1);
+		if (avalon2_send_pkgs(avalon2, &pkg))
+			return;
+	}
+
 		coinbase_len_prehash = pool->nonce2_offset - (pool->nonce2_offset % SHA256_BLOCK_SIZE);
 		coinbase_len_posthash = pool->coinbase_len - coinbase_len_prehash;
 		sha256_prehash(pool->coinbase, coinbase_len_prehash, coinbase_prehash);
@@ -918,7 +943,10 @@ static void avalon2_update(struct cgpu_info *avalon2)
 	memcpy(send_pkg.data + 8, &tmp, 4);
 
 	/* Configure the nonce2 offset and range */
-	range = 0xffffffff / (total_devices + 1);
+	if(pool->n2size == 3)
+		range = 0xffffff / (total_devices + 1);
+	else
+		range = 0xffffffff / (total_devices + 1);
 	start = range * (avalon2->device_id + 1);
 
 	tmp = be32toh(start);
diff --git a/driver-avalon2.h b/driver-avalon2.h
index 812b4dc..827c77d 100644
--- a/driver-avalon2.h
+++ b/driver-avalon2.h
@@ -75,6 +75,7 @@
 #define AVA2_P_STATUS		24
 #define AVA2_P_ACKDETECT	25
 #define AVA2_P_TEST_RET		26
+#define AVA2_P_LONGCOINBASE	27
 /* Avalon2 protocol package type */
 
 /* Avalon2/3 firmware prefix */
diff --git a/util.c b/util.c
index ea238a2..8b93578 100644
--- a/util.c
+++ b/util.c
@@ -1432,7 +1432,13 @@ bool extract_sockaddr(char *url, char **sockaddr_url, char **sockaddr_port)
 
 	if (url_len < 1)
 		return false;
-
+	
+	/* Get rid of the [] */
+	if (ipv6_begin && ipv6_end && ipv6_end > ipv6_begin){
+		url_len -= 2;
+		url_begin++;
+	}
+	
 	snprintf(url_address, 254, "%.*s", url_len, url_begin);
 
 	if (port_len) {
