From 7677a4507b1c9d024a5887c43c6136d82f17001a Mon Sep 17 00:00:00 2001
From: Xiangfu <xiangfu@openmobilefree.net>
Date: Fri, 8 Nov 2013 12:06:51 +0800
Subject: [PATCH 2/2] fix the temp_history_count for avalon 55nm prototype

---
 driver-avalon.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/driver-avalon.c b/driver-avalon.c
index 6c6f692..b6fe12b 100644
--- a/driver-avalon.c
+++ b/driver-avalon.c
@@ -859,7 +859,8 @@ static bool avalon_detect_one(libusb_device *dev, struct usb_find_devices *found
 	info->fan_pwm = AVALON_DEFAULT_FAN_MIN_PWM;
 	info->temp_max = 0;
 	/* This is for check the temp/fan every 3~4s */
-	info->temp_history_count = (4 / (float)((float)info->timeout * ((float)1.67/0x32))) + 1;
+	info->temp_history_count =
+		(4 / (float)((float)info->timeout * (110 / info->asic) * ((float)1.67/0x32))) + 1;
 	if (info->temp_history_count <= 0)
 		info->temp_history_count = 1;
 
-- 
1.8.1.2

