From 4f2fdad83cce55a5e0f4b75c6d8430ad62d881f9 Mon Sep 17 00:00:00 2001
From: Jimmy Situ <web@jimmystone.cn>
Date: Sun, 12 Apr 2015 17:17:26 +0800
Subject: [PATCH] atomic_pointer support on mips

---
 src/leveldb/port/atomic_pointer.h | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/leveldb/port/atomic_pointer.h b/src/leveldb/port/atomic_pointer.h
index a9866b2..3b1b109 100644
--- a/src/leveldb/port/atomic_pointer.h
+++ b/src/leveldb/port/atomic_pointer.h
@@ -38,6 +38,8 @@
 #define ARCH_CPU_ARM_FAMILY 1
 #elif defined(__ppc__) || defined(__powerpc__) || defined(__powerpc64__)
 #define ARCH_CPU_PPC_FAMILY 1
+#elif defined(__mips__) || defined(__mips64__)
+#define ARCH_CPU_MIPS_FAMILY 1
 #endif
 
 namespace leveldb {
@@ -102,6 +104,14 @@ inline void MemoryBarrier() {
 }
 #define LEVELDB_HAVE_MEMORY_BARRIER
 
+// MIPS
+#elif defined(ARCH_CPU_MIPS_FAMILY)
+
+inline void MemoryBarrier() {
+  __asm__ __volatile__("" : : : "memory");
+}
+#define LEVELDB_HAVE_MEMORY_BARRIER
+
 #endif
 
 // AtomicPointer built using platform-specific MemoryBarrier()
@@ -217,6 +227,7 @@ class AtomicPointer {
 #undef ARCH_CPU_X86_FAMILY
 #undef ARCH_CPU_ARM_FAMILY
 #undef ARCH_CPU_PPC_FAMILY
+#undef ARCH_CPU_MIPS_FAMILY
 
 }  // namespace port
 }  // namespace leveldb
-- 
1.9.1

